doc/api.rst

```diff
@@ -0,0 +1,234 @@
+of a given T point have the same array (*i*, *j*) indices. In
+the diagram below, the image on the left corresponds to choosing those
+points to the South and West of a T point to have the same (*i*, *j*)
+index. That on the right corresponds to choosing those points to the
+North and East of the T point (this is the offset scheme used in the
+NEMO ocean model):
+
+.. image:: grid_offset_choices.png
+
+The GOcean 1.0 API supports these two different offset schemes, which
+we term ``GO_OFFSET_SW`` and ``GO_OFFSET_NE``.
+
+Note that the constructor does not specify the extent of the model
+grid. This is because this information is normally obtained by reading
+a file (a namelist file, a netcdf file etc.) which is specific to an
+application.  Once this information has been obtained, a second
+routine, ``grid_init``, is provided with which to 'load' a grid object
+with state. This is discussed below.
+
+.. _gocean1.0-grid-init:
+
+The ``grid_init`` Routine
+#########################
+
+Once an application has determined the details of the model
+configuration, it must use this information to populate the grid
+object. This is done via a call to the ``grid_init`` subroutine::
+
+  subroutine grid_init(grid, m, n, dxarg, dyarg, tmask)
+    !> The grid object to configure
+    type(grid_type), intent(inout) :: grid
+    !> Dimensions of the model grid
+    integer,         intent(in)    :: m, n
+    !> The (constant) grid spacing in x and y (m)
+    real(wp),        intent(in)    :: dxarg, dyarg
+    !> Optional T-point mask specifying whether each grid point is
+    !! wet (1), dry (0) or external (-1).
+    integer, dimension(m,n), intent(in), optional :: tmask
+
+If no T-mask is supplied then this routine configures the grid
+appropriately for an all-wet domain with periodic boundary conditions
+in both the *x*- and *y*-dimensions. It should also be noted that
+currently only grids with constant resolution in *x* and *y* are
+supported by this routine.
+
+.. _gocean1.0-fields:
+
+Fields
+++++++
+
+Once a model has a grid defined it will require one or more
+fields. dl_esm_inf contains a ``field_mod`` module which defines an
+``r2d_field`` type (real, 2-dimensional field) and associated
+constructor::
+
+  use field_mod
+  ...
+  !> Current ('now') sea-surface height at different grid points
+  type(r2d_field) :: sshn_u_fld, sshn_v_fld, sshn_t_fld
+  ...
+
+  ! Sea-surface height now (current time step)
+  sshn_u = r2d_field(model_grid, GO_U_POINTS)
+  sshn_v = r2d_field(model_grid, GO_V_POINTS)
+  sshn_t = r2d_field(model_grid, GO_T_POINTS)
+
+The constructor takes two arguments:
+
+ 1. The grid on which the field exists
+ 2. The type of grid point at which the field is defined
+    (``GO_U_POINTS``, ``GO_V_POINTS``, ``GO_T_POINTS`` or ``GO_F_POINTS``)
+
+Note that the grid object need not have been fully configured (by a
+call to ``grid_init`` for instance) before it is passed into this
+constructor.
```

Comment:
I don't think this statement is true: the r2d_field_constructor accesses grid%nx, which as far as I can tell is only initialised after grid_init is called:
```
    upper_x_bound = self%grid%nx
    upper_y_bound = self%grid%ny
```
(lines 226 of field_mod.f90)

Reply:
Oh yes, that's true. I've altered the text accordingly.

Reply:
Great!